tests/fixtures/search.py

```diff
@@ -0,0 +1,293 @@
+    a type of search by changing analyzers or queries, but search needs to be tested manually
+    to ensure that it works well overall, with a realistic index.
+    """
+
+    SIMPLIFIED_TEST_ELASTICSEARCH = os.environ.get(
+        "SIMPLIFIED_TEST_ELASTICSEARCH", "http://localhost:9200"
+    )
+
+    indexes: List[Any]
+    integration: ExternalIntegration
+    search: Optional[SearchClientForTesting]
+    transaction: DatabaseTransactionFixture
+
+    @classmethod
+    def create(cls, data: DatabaseTransactionFixture) -> "ExternalSearchFixture":
+        fixture = ExternalSearchFixture()
+        fixture.transaction = data
+        fixture.indexes = []
+        fixture.integration = data.external_integration(
+            ExternalIntegration.ELASTICSEARCH,
+            goal=ExternalIntegration.SEARCH_GOAL,
+            url=ExternalSearchFixture.SIMPLIFIED_TEST_ELASTICSEARCH,
+            settings={
+                ExternalSearchIndex.WORKS_INDEX_PREFIX_KEY: "test_index",
+                ExternalSearchIndex.TEST_SEARCH_TERM_KEY: "test_search_term",
+            },
+        )
+
+        try:
+            fixture.search = SearchClientForTesting(data.session())
+        except Exception as e:
+            fixture.search = None
+            logging.error(
+                "Unable to set up elasticsearch index, search tests will be skipped.",
+                exc_info=e,
+            )
+        return fixture
+
+    def close(self):
+        if self.search:
+            # Delete the works_index, which is almost always created.
+            if self.search.works_index:
+                self.search.indices.delete(self.search.works_index, ignore=[404])
+            # Delete any other indexes created over the course of the test.
+            for index in self.indexes:
+                self.search.indices.delete(index, ignore=[404])
+            ExternalSearchIndex.reset()
+
+    def setup_index(self, new_index):
+        """Create an index and register it to be destroyed during teardown."""
+        self.search.setup_index(new_index=new_index)
+        self.indexes.append(new_index)
+
+    def default_work(self, *args, **kwargs):
+        """Convenience method to create a work with a license pool in the default collection."""
+        work = self.transaction.work(
+            *args,
+            with_license_pool=True,
+            collection=self.transaction.default_collection(),
+            **kwargs
+        )
+        work.set_presentation_ready()
+        return work
+
+
+@pytest.fixture(scope="function")
+def external_search_fixture(
+    database_transaction: DatabaseTransactionFixture,
+) -> Iterable[ExternalSearchFixture]:
+    """Ask for an external search system."""
+    """Note: You probably want EndToEndSearchFixture instead."""
+    data = ExternalSearchFixture.create(database_transaction)
+    yield data
+    data.close()
+
+
+T = TypeVar("T")
+
+
+class EndToEndSearchFixture(Generic[T]):
```

Comment:
Why do we have this generic? I thought only `Work` was supported

Reply:
The way the previous code worked was that test classes extended an `EndToEndSearchTest` class, and defined a `populate_works()` method that assigned a completely test-specific set of attributes to `self`.

Replacing the `EndToEndSearchTest` class with a fixture meant that there could no longer be a `populate_works()` method that tests would implement, so we needed a different place for tests to store the attributes they needed. A first attempt at this produced a fixture that looked like this:

```
class EndToEndSearchFixture:
  def populate(f: Callable[EndToEndSearchFixture, ...]):
    f(self)
```

Tests would then call the `populate` method and assign fields to the fixture:

```
def _populate(data: EndToEndSearchFixture):
  data.something = "..."
  data.something_else = "..."
  data.something_else2 = "..."

def test_something(self, end_to_end_search_fixture: EndToEndSearchFixture):
  end_to_end_search_fixture.populate(self._populate_data);

  do_something_with_data(end_to_end_search_fixture.something, ...)
```

The problem with this is that it loses type-safety: The code in `test_something` would tell the fixture to call `self._populate_data`, and this would create new attributes on the fixture but the type checker would be completely unaware that the created attributes existed. The approach I then took was to have the `EndToEndSearchFixture` explicitly state the type of test-specific user data with type parameter `T`, and have the called `_populate_data` method be responsible for returning a value of type `T` that the tests can then use in a type-safe way.

It's not the way I would have written these tests from a blank canvas, but it was the least invasive way to convert the existing tests.

Reply:
The `self.test_data` property is used, but it's only used inside the test methods. It's kind of an odd situation, because the tests are structured such that the fixture is initialized, and then has to have test-specific data assigned, and then data has to be indexed, and _then_ the test runs. In order to preserve type-safety between all of these steps, the fixture itself has to carry the type parameter.

Reply:
Rather than accessing the works through the fixture, why not just have `self.data = TestExternalSearchWithWorksData()` in `TestExternalSearchWithWorks._populate_works`

Reply:
This would yield the same kinds of mutable state problems that the fixtures are intended to prevent: Test methods have to be disciplined about not accidentally changing state in a way that could affect other test methods in the same class. Having the fixture carry the state guarantees this can't happen (because the fixture is always destroyed and recreated for each test method).

Reply:
The test class is also destroyed and recreated between each test method. 
Now that I'm thinking about the lifecycle, does there need to be a callback function to create the works at all? 
The test method can just create the works up front within itself, and then ask the fixture to run the coverage provider indexing.
Since the coverage provider only cares about the works in the DB, the fixture doesn't need to know about what works have been created.

Reply:
I was basing my approach on this comment made back in February:

https://github.com/ThePalaceProject/circulation/pull/178#discussion_r800983153

But I now realize that's referring only to class variables and not instance variables.

We could get rid of the callback if the tests only use instance variables.

Reply:
@jonathangreen What do you think? ^^

Reply:
I didn't do a very good job describing class variables vs instance variables in that comment did I 😅 

I think it makes sense for the tests to create the works as they see fit, then ask the fixture for indexing. Eliminating the callback seems like it would simplify things and provide better separation of concerns.

Reply:
Thankfully, I know the difference, I just didn't realize that `pytest` would actually recreate class instances each time. I'll get rid of the callback.

Reply:
Turns out we got to go one better: None of the tests were dependent on using `self`, so the callback has been removed and everything has been reduced to local variables inside test methods.